include/nbl/asset/filters/CNormalMapToDerivativeFilter.h

```diff
@@ -0,0 +1,275 @@
+		{
+			if (!state)
+				return false;
+
+			return true;
+		}
+};
+
+//! Convert Normal Map to Derivative Normal Map
+/*
+	
+*/
+
+class CNormalMapToDerivativeFilter : public CMatchedSizeInOutImageFilterCommon, public CNormalMapToDerivativeFilterBase
+{
+	public:
+		virtual ~CNormalMapToDerivativeFilter() {}
+
+		class CStateBase : public CMatchedSizeInOutImageFilterCommon::state_type, public CNormalMapToDerivativeFilterBase::CNormalMapToDerivativeStateBase
+		{ 
+			public:
+				CStateBase() = default;
+				virtual ~CStateBase() = default;
+
+			private:
+
+				void resetLayerScaleValues()
+				{
+					maxAbsLayerScaleValues.clear();
+				}
+
+				friend class CNormalMapToDerivativeFilter;
+		};
+		using state_type = CStateBase; //!< full combined state
+
+		static inline bool validate(state_type* state)
+		{
+			if (!CMatchedSizeInOutImageFilterCommon::validate(state))
+				return false;
+
+			if (!CNormalMapToDerivativeFilterBase::validate(state))
+				return false;
+			
+			const ICPUImage::SCreationParams& inParams = state->inImage->getCreationParameters();
+			const ICPUImage::SCreationParams& outParams = state->outImage->getCreationParameters();
+			const auto inFormat = inParams.format;
+			const auto outFormat = outParams.format;
+
+			if (outFormat != asset::EF_R8G8_SNORM)
+				return false;
+
+			if (state->scratchMemoryByteSize < state_type::getRequiredScratchByteSize(state->extent))
+				return false;
+
+			if (asset::getFormatChannelCount(inFormat) < 3 )
+				return false;
+
+			return true;
+		}
+
+		static inline bool execute(state_type* state)
+		{
+			if (!validate(state))
+				return false;
+
+			state->resetLayerScaleValues();
+
+			auto checkFormat = state->inImage->getCreationParameters().format;
+			if (isIntegerFormat(checkFormat))
+				return executeInterprated(state, reinterpret_cast<uint64_t*>(state->scratchMemory));
+			else
+				return executeInterprated(state, reinterpret_cast<double*>(state->scratchMemory));
+		}	
+
+	private:
+
+		template<typename decodeType> //!< double or uint64_t
+		static inline bool executeInterprated(state_type* state, decodeType* scratchMemory)
+		{
+			const asset::E_FORMAT inFormat = state->inImage->getCreationParameters().format;
+			const auto inTexelByteSize = asset::getTexelOrBlockBytesize(inFormat);
+			const auto currentChannelCount = asset::getFormatChannelCount(inFormat);
+			const auto arrayLayers = state->inImage->getCreationParameters().arrayLayers;
+			static constexpr auto maxChannels = 4u;
+
+			#ifdef _NBL_DEBUG
+			memset(scratchMemory, 0, state->scratchMemoryByteSize);
+			#endif // _NBL_DEBUG
+
+			const core::vector3du32_SIMD scratchByteStrides = TexelBlockInfo(asset::E_FORMAT::EF_R64G64B64A64_SFLOAT).convert3DTexelStridesTo1DByteStrides(state->extentLayerCount);
+			const auto scratchTexelByteSize = scratchByteStrides[0];
+
+			// I wonder if we should let somebody pass through more than 1 layer, though I find it cool
```

Comment:
and this @Devsh-Graphics-Programming/seniors

Reply:
treat each layer separately (each layer should have its own scale factor), but yes multiple layers might be the case if we ever get a loader where an array texture is used for multiple geometries with 1 layer per meshbuffer